client/src/state.rs

```diff
@@ -0,0 +1,307 @@
+pub struct StateApi {
+    ctx: Arc<Context>,
+}
+
+impl StateApi {
+    pub(crate) fn new(ctx: Arc<Context>) -> StateApi {
+        StateApi { ctx }
+    }
+
+    /// Returns the public key of the signer.
+    pub fn pubkey(&self) -> Result<VerifyingKey> {
+        self.ctx.pubkey().cloned()
+    }
+
+    /// Returns the address of signer.
+    pub fn account_address(&self) -> Result<AccAddress> {
+        let pubkey = self.ctx.pubkey()?.to_owned();
+        Ok(AccAddress::new(pubkey.into()))
+    }
+
+    /// Retrieves the Celestia coin balance for the signer.
+    pub async fn balance(&self) -> Result<u64> {
+        let address = self.account_address()?;
+        self.balance_for_address(address).await
+    }
+
+    /// Retrieves the Celestia coin balance for the given address.
+    ///
+    /// # Notes
+    ///
+    /// This is the only method of [`StateApi`] that fallbacks to bridge node
+    /// if consensus node is not set in [`Client`].
+    pub async fn balance_for_address(&self, address: AccAddress) -> Result<u64> {
+        let address = Address::AccAddress(address);
+
+        let grpc = match self.ctx.grpc() {
+            Ok(grpc) => grpc,
+            Err(_) => {
+                return Ok(self
+                    .ctx
+                    .rpc
+                    .state_balance_for_address(&address)
+                    .await?
+                    .amount());
+            }
+        };
+
+        // TODO: Verify balance with AbciQuery is ready.
+        let coin = grpc.get_balance(&address, "utia").await?;
+
+        Ok(coin.amount())
+    }
+
+    /// Submit given message to celestia network.
+    ///
+    /// When no gas price is specified through config, it will automatically
+    /// handle updating client's gas price when consensus updates minimal
+    /// gas price.
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use celestia_client::{Client, Result};
+    /// # use celestia_client::tx::TxConfig;
+    /// # const RPC_URL: &str = "http://localhost:26658";
+    /// # const GRPC_URL : &str = "http://localhost:19090";
+    /// # async fn docs() -> Result<()> {
+    /// use celestia_proto::cosmos::bank::v1beta1::MsgSend;
+    /// use celestia_types::state::{Address, Coin};
+    ///
+    /// let client = Client::builder()
+    ///     .rpc_url(RPC_URL)
+    ///     .grpc_url(GRPC_URL)
+    ///     .plaintext_private_key("...")?
+    ///     .build()
+    ///     .await?;
+    ///
+    /// let msg = MsgSend {
+    ///     from_address: client.state().account_address()?.to_string(),
+    ///     to_address: "celestia169s50psyj2f4la9a2235329xz7rk6c53zhw9mm".to_string(),
+    ///     amount: vec![Coin::utia(12345).into()],
+    /// };
+    ///
+    /// client
+    ///     .state()
+    ///     .submit_message(msg, TxConfig::default())
+    ///     .await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn submit_message<M>(&self, message: M, cfg: TxConfig) -> Result<TxInfo>
```

Comment:
Should I move this to `Client`? Go celestia-node does not expose such functionality. Also I may expose `broadcast_tx` later on.